src/frontend/constraint_system.rs

```diff
@@ -0,0 +1,432 @@
+  /// in testing contexts in order to derive a unique name for this variable in the current
+  /// namespace.
+  fn alloc<F, A, AR>(&mut self, annotation: A, f: F) -> Result<Variable, SynthesisError>
+  where
+    F: FnOnce() -> Result<Scalar, SynthesisError>,
+    A: FnOnce() -> AR,
+    AR: Into<String>;
+
+  /// Allocate a public variable in the constraint system. The provided function is used to
+  /// determine the assignment of the variable.
+  fn alloc_input<F, A, AR>(&mut self, annotation: A, f: F) -> Result<Variable, SynthesisError>
+  where
+    F: FnOnce() -> Result<Scalar, SynthesisError>,
+    A: FnOnce() -> AR,
+    AR: Into<String>;
+
+  /// Enforce that `A` * `B` = `C`. The `annotation` function is invoked in testing contexts
+  /// in order to derive a unique name for the constraint in the current namespace.
+  fn enforce<A, AR, LA, LB, LC>(&mut self, annotation: A, a: LA, b: LB, c: LC)
+  where
+    A: FnOnce() -> AR,
+    AR: Into<String>,
+    LA: FnOnce(LinearCombination<Scalar>) -> LinearCombination<Scalar>,
+    LB: FnOnce(LinearCombination<Scalar>) -> LinearCombination<Scalar>,
+    LC: FnOnce(LinearCombination<Scalar>) -> LinearCombination<Scalar>;
+
+  /// Create a new (sub)namespace and enter into it. Not intended
+  /// for downstream use; use `namespace` instead.
+  fn push_namespace<NR, N>(&mut self, name_fn: N)
+  where
+    NR: Into<String>,
+    N: FnOnce() -> NR;
+
+  /// Exit out of the existing namespace. Not intended for
+  /// downstream use; use `namespace` instead.
+  fn pop_namespace(&mut self);
+
+  /// Gets the "root" constraint system, bypassing the namespacing.
+  /// Not intended for downstream use; use `namespace` instead.
+  fn get_root(&mut self) -> &mut Self::Root;
+
+  /// Begin a namespace for this constraint system.
+  fn namespace<NR, N>(&mut self, name_fn: N) -> Namespace<'_, Scalar, Self::Root>
+  where
+    NR: Into<String>,
+    N: FnOnce() -> NR,
+  {
+    self.get_root().push_namespace(name_fn);
+
+    Namespace(self.get_root(), Default::default())
+  }
+
+  /// Most implementations of ConstraintSystem are not 'extensible': they won't implement a specialized
+  /// version of `extend` and should therefore also keep the default implementation of `is_extensible`
+  /// so callers which optionally make use of `extend` can know to avoid relying on it when unimplemented.
+  fn is_extensible() -> bool {
+    false
+  }
+
+  /// Extend concatenates thew  `other` constraint systems to the receiver, modifying the receiver, whose
+  /// inputs, allocated variables, and constraints will precede those of the `other` constraint system.
+  /// The primary use case for this is parallel synthesis of circuits which can be decomposed into
+  /// entirely independent sub-circuits. Each can be synthesized in its own thread, then the
+  /// original `ConstraintSystem` can be extended with each, in the same order they would have
+  /// been synthesized sequentially.
+  fn extend(&mut self, _other: &Self) {
```

Comment:
Nit: This method seems unimplemented. Is it worth cutting out unimplemented methods to keep the codebase simple?

Reply:
It seems like there are some places this actually gets used. Will look more.